File: scripts/orblv_by_gencode_ribotie.R
library(ggplot2)
library(dplyr)

df <- read.delim(
  "nextflow_results/quality/minlen_quality_metrics.tsv",
  stringsAsFactors = FALSE
) |>
  filter(in_gencode_ribotie %in% c("True", "False")) |>
  mutate(
    ORBLv = as.numeric(ORBLv),
    in_gencode_ribotie = factor(
      in_gencode_ribotie,
      levels = c("True", "False"),
      labels = c("In GENCODE RiboTIE", "Not in GENCODE RiboTIE")
    )
  ) |>
  filter(!is.na(ORBLv))

counts <- df |> count(in_gencode_ribotie)
label_map <- setNames(
  sprintf("%s\n(n = %s)", counts$in_gencode_ribotie, format(counts$n, big.mark = ",")),
  as.character(counts$in_gencode_ribotie)
)
df$group_label <- label_map[as.character(df$in_gencode_ribotie)]
df$group_label <- factor(df$group_label, levels = label_map)

wt <- wilcox.test(
  ORBLv ~ in_gencode_ribotie,
  data = df,
  exact = FALSE
)
p_str <- if (wt$p.value < 1e-300) "p < 1e-300" else sprintf("p = %.2e", wt$p.value)

medians <- df |> group_by(in_gencode_ribotie) |> summarise(med = median(ORBLv))

p <- ggplot(df, aes(x = ORBLv, colour = group_label, fill = group_label)) +
  geom_density(alpha = 0.35, linewidth = 0.8) +
  geom_vline(
    data = medians,
    aes(xintercept = med, colour = factor(
      in_gencode_ribotie,
      levels = c("In GENCODE RiboTIE", "Not in GENCODE RiboTIE"),
      labels = label_map
    )),
    linetype = "dashed", linewidth = 0.6
  ) +
  scale_colour_manual(values = c("In GENCODE RiboTIE\n(n = 22,626)" = "#2196F3",
                                  "Not in GENCODE RiboTIE\n(n = 28,399)" = "#E53935")) +
  scale_fill_manual(values   = c("In GENCODE RiboTIE\n(n = 22,626)" = "#2196F3",
                                  "Not in GENCODE RiboTIE\n(n = 28,399)" = "#E53935")) +
  annotate("text", x = 0.98, y = Inf, label = paste0("Wilcoxon rank-sum\n", p_str),
           hjust = 1, vjust = 1.4, size = 3.2, colour = "grey30") +
  labs(
    x = "ORBLv score",
    y = "Density",
    title = "ORBLv score distribution by GENCODE RiboTIE membership",
    colour = NULL, fill = NULL
  ) +
  theme_classic(base_size = 12) +
  theme(legend.position = "top")

out <- "nextflow_results/quality/orblv_by_gencode_ribotie.png"
ggsave(out, p, width = 7, height = 4.5, dpi = 150)
cat("Saved", out, "\n")

df |>
  group_by(in_gencode_ribotie) |>
  summarise(n = n(), median = median(ORBLv), mean = mean(ORBLv)) |>
  print()
cat("Wilcoxon W =", wt$statistic, ",", p_str, "\n")

File: scripts/orblv_by_gencode_ribotie.py
#!/usr/bin/env python3
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D

df = pd.read_csv(
    "nextflow_results/quality/minlen_quality_metrics.tsv",
    sep="\t",
    usecols=["ORBLv", "in_gencode_ribotie"],
)
df = df[df["in_gencode_ribotie"].isin([True, False, "True", "False"])].copy()
df["ORBLv"] = pd.to_numeric(df["ORBLv"], errors="coerce")
df["in_gencode_ribotie"] = df["in_gencode_ribotie"].astype(str)
df = df.dropna(subset=["ORBLv"])

groups = {
    "True":  df[df["in_gencode_ribotie"] == "True"]["ORBLv"].values,
    "False": df[df["in_gencode_ribotie"] == "False"]["ORBLv"].values,
}

colors = {"True": "#2196F3", "False": "#E53935"}
labels = {
    "True":  f"In GENCODE RiboTIE (n={len(groups['True']):,})",
    "False": f"Not in GENCODE RiboTIE (n={len(groups['False']):,})",
}

fig, ax = plt.subplots(figsize=(7, 4.5))

x = np.linspace(0, 1, 500)
for key in ("True", "False"):
    vals = groups[key]
    # KDE via numpy histogram + smoothing
    counts, edges = np.histogram(vals, bins=100, range=(0, 1), density=True)
    centers = (edges[:-1] + edges[1:]) / 2
    ax.fill_between(centers, counts, alpha=0.4, color=colors[key], linewidth=0)
    # Smooth KDE line using a simple Gaussian kernel via convolution
    sigma_bins = 3
    kernel = np.exp(-0.5 * (np.arange(-3*sigma_bins, 3*sigma_bins+1) / sigma_bins) ** 2)
    kernel /= kernel.sum()
    smoothed = np.convolve(counts, kernel, mode="same")
    ax.plot(centers, smoothed, color=colors[key], linewidth=1.8, label=labels[key])
    med = np.median(vals)
    ax.axvline(med, color=colors[key], linestyle="--", linewidth=1.0, alpha=0.8)

ax.set_xlabel("ORBLv score", fontsize=12)
ax.set_ylabel("Density", fontsize=12)
ax.set_title("ORBLv score distribution by GENCODE RiboTIE membership", fontsize=12)
ax.set_xlim(0, 1)
ax.legend(fontsize=10, frameon=False)
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)

plt.tight_layout()
out = "nextflow_results/quality/orblv_by_gencode_ribotie.png"
plt.savefig(out, dpi=150)
print(f"Saved {out}")

for key in ("True", "False"):
    vals = groups[key]
    print(f"{labels[key]}: median={np.median(vals):.4f}  mean={np.mean(vals):.4f}")
